pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java

```diff
@@ -0,0 +1,360 @@
+        if (cardinality > _cardinalityThreshold) {
+          // Do not append inverted index if already exists
+          if (!invertedIndexColumns.contains(columnName)) {
+            invertedIndexColumns.add(columnName);
+          }
+          LOGGER.info("Table: {}, add inverted index to column {} with cardinality: {}", tableName, columnName,
+              cardinality);
+        } else {
+          LOGGER.info("Table: {}, skip adding inverted index to column {} with cardinality: {}", tableName, columnName,
+              cardinality);
+          break;
+        }
+      }
+
+      // Update indexing config
+      if (!invertedIndexColumns.isEmpty()) {
+        indexingConfig.setAutoGeneratedInvertedIndex(true);
+        if (updateIndexConfig(tableName, tableConfig)) {
+          LOGGER.info("Table: {}, added inverted index to columns: {}", tableName, invertedIndexColumns);
+        } else {
+          LOGGER.error("Table: {}, failed to add inverted index to columns: {}", tableName, invertedIndexColumns);
+        }
+      } else {
+        if (autoGeneratedInvertedIndex) {
+          assert _mode == Mode.REFRESH;
+
+          // Remove existing auto-generated inverted index because no column matches all the conditions
+          indexingConfig.setAutoGeneratedInvertedIndex(false);
+          if (updateIndexConfig(tableName, tableConfig)) {
+            LOGGER.info("Table: {}, removed auto-generated inverted index", tableName);
+          } else {
+            LOGGER.error("Table: {}, failed to remove auto-generated inverted index", tableName);
+          }
+        }
+      }
+    }
+  }
+
+  private AbstractTableConfig getTableConfig(String tableName)
+      throws Exception {
+    return AbstractTableConfig.fromZnRecord(_propertyStore.get(TABLE_CONFIG_PATH + tableName, null, 0));
+  }
+
+  private Schema getTableSchema(String tableName)
+      throws Exception {
+    ZNRecord znRecord = _propertyStore.get(SCHEMA_PATH + tableName, null, 0);
+    if (znRecord == null) {
+      return null;
+    } else {
+      return SchemaUtils.fromZNRecord(znRecord);
+    }
+  }
+
+  private JSONObject sendQuery(String query)
+      throws Exception {
+    URLConnection urlConnection = new URL("http://" + _brokerAddress + "/query").openConnection();
+    urlConnection.setDoOutput(true);
+
+    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(urlConnection.getOutputStream(), "UTF-8"));
+    writer.write(new JSONObject().put("pql", query).toString());
+    writer.flush();
+
+    BufferedReader reader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "UTF-8"));
+    return new JSONObject(reader.readLine());
+  }
+
+  private boolean updateIndexConfig(String tableName, AbstractTableConfig tableConfig)
+      throws Exception {
+    HttpURLConnection httpURLConnection = (HttpURLConnection) new URL(
+        "http://" + _controllerAddress + "/tables/" + tableName + "/indexingConfigs").openConnection();
```

Comment:
Use `ControllerRequestURLBuilder` to build REST API URLs. In this case, you'll need to add the one to upload the indexing configs.